### PR TITLE
HAI-2065 Implement file upload component

### PR DESCRIPTION
**Pull request by markohaarni**
# Description

File upload component uses HDS FileInput internally to handle user adding files from their machine and immediately sends added files to backend and shows a loading spinner and loading text while upload is in process.

Listing of files and deleting them as well as displaying error messages are done in separate tasks.

## Type of change

- [ ] Bug fix
- [ ] New feature
- [x] Other

# Instructions for testing

Only way to manually test this is to replace FileInput component locally in src/domain/johtoselvitys/Attachments.tsx with this new component, for example:
```
function handleFileUpload(uploading: boolean) {
  if (!uploading) {
    queryClient.invalidateQueries('attachments');
  }
}

<FileUpload
  id="cable-report-file-upload"
  accept=".pdf,.jpg,.jpeg,.png,.dgn,.dwg,.docx,.txt,.gt"
  maxSize={104857600}
  dragAndDrop
  multiple
  existingAttachments={existingAttachments}
  uploadFunction={(file) =>
    uploadAttachment({
      applicationId: getValues('id')!,
      attachmentType: 'MUU',
      file,
    })
  }
  onUpload={handleFileUpload}
/>
```

# Checklist:

- [x] I have written new tests (if applicable)
- [x] I have ran the tests myself (if applicable)
- [ ] I have made necessary changes to the documentation, link to confluence
      or other location:

# Other relevant info

Please describe here if there is e.g. some requirements for this change or
other info that the tester/user needs to know.

### CODE REVIEW

File: src/common/components/fileUpload/utils.ts [hunk 1]
```diff
@@ -0,0 +1,15 @@
+import { AttachmentMetadata } from '../../types/attachment';
+
+// Filter out duplicate files based on file name
+export function removeDuplicateAttachments<T extends AttachmentMetadata>(
+  files: File[],
+  attachments: T[] | undefined,
```
Comment (pitkni) [inline src/common/components/fileUpload/utils.ts:6]: Would it make sense to rename these to something like existingAttachments and newFiles?

File: src/common/components/fileUpload/utils.ts [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+): [File[], File[]] {
+  const duplicateFiles = files.filter(
+    (file) => attachments?.some((attachment) => attachment.fileName === file.name),
+  );
+  const newFiles = files.filter(
+    (file) => attachments?.every((attachment) => attachment.fileName !== file.name),
+  );
+  return [newFiles, duplicateFiles];
```
[the remark this answers is not in the document]
Comment (markohaarni) [reply src/common/components/fileUpload/utils.ts:14]: Yes, it's needed for generating the error messages for duplicate files.

File: src/common/components/fileUpload/utils.ts (hunk 1 above)
Comment (markohaarni) [reply src/common/components/fileUpload/utils.ts]: Good point. Renamed those.

File: src/common/components/fileUpload/FileUpload.tsx [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+    if (ref.current) {
+      ref.current.ondrop = dropHandler;
+    }
+  }, []);
+
+  return { ref, files };
+}
+
+function SuccessNotification({
+  successfulCount,
```
Comment (pitkni) [inline src/common/components/fileUpload/FileUpload.tsx:34]: There's a couple of sonar code smells (props should be read only here), should they be fixed?

Comment (markohaarni) [reply src/common/components/fileUpload/FileUpload.tsx:34]: I don't know if it's necessary to fix them as eslint throws error if there is parameter reassigning.

Comment (markohaarni) [reply src/common/components/fileUpload/FileUpload.tsx:34]: Fixed them anyway.

**Review by pitkni — APPROVED** [review]

**Pull request merged**